### PR TITLE
Add reservedTo field to Node

### PR DESCRIPTION
**Pull request by bratseth**

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java [hunk 1]
```diff
@@ -265,26 +273,32 @@ public Node allocate(ApplicationId owner, ClusterMembership membership, NodeReso
...
 
     /** Returns a copy of this node with the parent hostname assigned to the given value. */
     public Node withParentHostname(String parentHostname) {
         return new Node(id, ipConfig, hostname, Optional.of(parentHostname), flavor, status, state,
-                        allocation, history, type, reports, modelName);
+                        allocation, history, type, reports, modelName, reservedTo);
+    }
+
+    /** Returns a copy of this node marked as reserved to the given tenant (or empty to remove reservation) */
+    public Node withReservedTo(Optional<TenantName> reservedTo) {
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java]: The convention is this class seems to be to have a `withReservedTo(TenantName)` and `withoutReservedTo()`

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/Activator.java [hunk 2]
```diff
@@ -90,6 +89,27 @@ private void activateNodes(ApplicationId application, Collection<HostSpec> hosts
...
+        List<String> reservedParents = nodes.stream()
+                                            .filter(node -> node.parentHostname().isPresent())
+                                            .map(node -> nodeRepository.getNode(node.parentHostname().get()))
+                                            .filter(parent -> parent.isPresent())
+                                            .filter(parent -> parent.get().reservedTo().isPresent())
+                                            .map(parent -> parent.get().hostname())
+                                            .collect(Collectors.toList());
+        if (reservedParents.isEmpty()) return;
+
+        try (Mutex lock = nodeRepository.lockUnallocated()) {
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/Activator.java]: This should lock the tenant host application

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodePrioritizer.java [hunk 3]
```diff
@@ -122,12 +122,14 @@ void addSurplusNodes(List<Node> surplusNodes) {
      *
      * @param exclusively whether the ready docker nodes should only be added on hosts that
      *                    already have nodes allocated to this tenant
+     * @param application the application we are adding nodes for
      */
-    void addNewDockerNodes(boolean exclusively) {
+    void addNewDockerNodes(boolean exclusively, ApplicationId application) {
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodePrioritizer.java]: Use `appId` in this class

**Review by freva — COMMENTED** [review]
Build failure:
```
[ERROR] /source/athenz-identity-provider-service/src/test/java/com/yahoo/vespa/hosted/athenz/instanceproviderservice/instanceconfirmation/InstanceValidatorTest.java:[226,29] method create in class com.yahoo.vespa.hosted.provision.Node cannot be applied to given types;
[ERROR]   required: java.lang.String,com.yahoo.vespa.hosted.provision.node.IP.Config,java.lang.String,java.util.Optional<java.lang.String>,java.util.Optional<java.lang.String>,com.yahoo.config.provision.Flavor,java.util.Optional<com.yahoo.config.provision.TenantName>,com.yahoo.config.provision.NodeType
[ERROR]   found: java.lang.String,com.yahoo.vespa.hosted.provision.node.IP.Config,java.lang.String,java.util.Optional<java.lang.Object>,java.util.Optional<java.lang.Object>,com.yahoo.config.provision.Flavor,com.yahoo.config.provision.NodeType
[ERROR]   reason: actual and formal argument lists differ in length
[ERROR] /source/athenz-identity-provider-service/src/test/java/com/yahoo/vespa/hosted/athenz/instanceproviderservice/identitydocument/IdentityDocumentGeneratorTest.java:[61,31] method create in class com.yahoo.vespa.hosted.provision.Node cannot be applied to given types;
[ERROR]   required: java.lang.String,com.yahoo.vespa.hosted.provision.node.IP.Config,java.lang.String,java.util.Optional<java.lang.String>,java.util.Optional<java.lang.String>,com.yahoo.config.provision.Flavor,java.util.Optional<com.yahoo.config.provision.TenantName>,com.yahoo.config.provision.NodeType
[ERROR]   found: java.lang.String,com.yahoo.vespa.hosted.provision.node.IP.Config,java.lang.String,java.util.Optional<java.lang.Object>,java.util.Optional<java.lang.Object>,com.yahoo.config.provision.Flavor,com.yahoo.config.provision.NodeType
[ERROR]   reason: actual and formal argument lists differ in length
```

**Comment by freva** [comment]
Also note there is some code in `hosted` that creates `Node`s and might need to be updated to pass the new field.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java [hunk 4]
```diff
@@ -41,61 +42,57 @@
...
     private final Reports reports;
     private final Optional<String> modelName;
+    private final Optional<TenantName> reservedTo;
 
     /** Record of the last event of each type happening to this node */
     private final History history;
 
     /** The current allocation of this node, if any */
     private final Optional<Allocation> allocation;
+    private Node retire;
```
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java]: Yes, thanks!

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java (hunk 1 above)
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java]: Done

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/Activator.java (hunk 2 above)
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/Activator.java]: That stuff has become pretty weird, and it doesn't seem that's how it is consistently done? Anyway; ok, done.

**Comment by bratseth** [comment]
PTAL @freva

**Review by freva — APPROVED** [review]

**Pull request merged**